cpp/src/arrow/array/builder_union.h

```diff
@@ -47,12 +47,23 @@ class ARROW_EXPORT DenseUnionBuilder : public ArrayBuilder {
   explicit DenseUnionBuilder(MemoryPool* pool,
                              const std::shared_ptr<DataType>& type = NULLPTR);
 
-  Status AppendNull() {
+  Status AppendNull() final {
     ARROW_RETURN_NOT_OK(types_builder_.Append(0));
     ARROW_RETURN_NOT_OK(offsets_builder_.Append(0));
     return AppendToBitmap(false);
   }
 
+  Status AppendNulls(int64_t length) final {
+    ARROW_RETURN_NOT_OK(types_builder_.Reserve(length));
+    ARROW_RETURN_NOT_OK(offsets_builder_.Reserve(length));
+    ARROW_RETURN_NOT_OK(Reserve(length));
+    for (int64_t i = 0; i < length; ++i) {
```

Comment:
Also memcpy worthy.

Reply:
Perhaps you mean `memset`. Though the compiler may be able to infer the optimization by itself.

Reply:
derp, I meant `memset` yes. I'm not sure if the compiler can elide and transform, it really depends on inlining, I also wouldn't expect this to work with all compilers.

Reply:
We would need either to break the builder abstraction or add specific zero-appending methods to numeric builders. I would rather keep the code as-is for now.